torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -3489,6 +3489,160 @@ Tensor i1e_backward(
+// hence
+// LU_grad = L_grad * 1_L + U_grad * 1_U (!!!)
+//
+// Using the definition of dX:
+// Tr(X_grad^H X) = Tr(-U^{-1} L^{-1} P^T B X_grad^H U^{-1} dU)
+//                + Tr(-L^{-1} P^T B X_grad^H U^{-1} L^{-1} dL)
+//                + Tr(X_grad^H U^{-1} L^{-1} P^T dB).
+// And we immediately get:
+// B_grad = [X_grad^H U^{-1} L^{-1} P^T]^H = [U^{-1} L^{-1} P^T]^H X_grad.
+// Let Z := L^{-1} P^T B X_grad^H U^{-1}, then
+// U_grad = [-U^{-1} Z]^H,
+// L_grad = [-Z L^{-1}]^H.
+// After inserting U_grad and L_grad into (!!!) we get the value for LU_grad.
+
+Tensor lu_solve_backward_lu_data(
+  const Tensor& grad,
+  const Tensor& self,
+  const Tensor& LU_data,
+  const Tensor& LU_pivots
+) {
+  Tensor P, L, U;
+  std::tie(P, L, U) = at::lu_unpack(LU_data, LU_pivots);
+
+  auto n = LU_data.size(-1);
+  auto nrhs = self.size(-1);
+
+  // X = -L^{-1} P^T B grad^H
+  auto X = -std::get<0>(at::triangular_solve(
+    (nrhs < n) ? P.transpose(-2, -1).matmul(self) : P.transpose(-2, -1),
+    L,
+    /*upper=*/false,
+    /*transpose=*/false,
+    /*unitriangular=*/true
+  ));
+  if (nrhs >= n) {
+    X = X.matmul(self);
+  }
+  X = X.matmul(grad.transpose(-2, -1).conj());
+
+  // X <- X U^{-1}
+  X = std::get<0>(at::triangular_solve(
+    X.transpose(-2, -1).conj(),
+    U.transpose(-2, -1).conj(),
+    /*upper=*/false,
+    /*transpose=*/false,
+    /*unitriangular=*/false
+  )).transpose(-2, -1).conj();
+
+  // U_grad = [U^{-1} X]^H
+  auto U_grad = std::get<0>(at::triangular_solve(
+    X,
+    U,
+    /*upper=*/true,
+    /*transpose=*/false,
+    /*unitriangular=*/false
+  )).transpose(-2, -1).conj();
+
+  // L_grad = L^{-H} X^H
+  auto L_grad = std::get<0>(at::triangular_solve(
+    X.transpose(-2, -1).conj(),
+    L.transpose(-2, -1).conj(),
+    /*upper=*/true,
+    /*transpose=*/false,
+    /*unitriangular=*/true
+  ));
+
+  // Suppose
+  // 1 = ones(n, n)
+  // 1_U = 1.tril()
+  // 1_L = 1 - 1_U (note zero diagonal),
+  // * - the elementwise product.
+  //
+  // LU_data could be represented as
+  // LU_data = L + U - I, so
+  // dLU_data = dL + dU.
+  // Since the diagonal of L is never explicitly exposed, it follows that
+  // diag(dL) = 0, and hence:
+  // dL = dLU_data * 1_L,
+  // dU = dLU_data * 1_U.
+  // These would imply:
+  // LU_data_grad = L_grad * 1_L + U_grad * 1_U
+
+  // we cannot kill the diagonal of L_grad in-place as this varible
+  // is needed in the triangular_solve's (double) backward,
+  // hence we need a tensor of diag(L_grad) that is going to be
+  // subtracted from L_grad
+  auto L_grad_diag_embed = at::diag_embed(L_grad.diagonal(0, -2, -1));
+
+  // LU_data_grad = L_grad * 1_L + U_grad * 1_U
+  auto LU_data_grad = L_grad.tril() - L_grad_diag_embed + U_grad.triu();
+  return LU_data_grad;
+}
+
+Tensor lu_solve_backward_self(
+  const Tensor& grad,
+  const Tensor& self,
+  const Tensor& LU_data,
+  const Tensor& LU_pivots
+) {
+  Tensor P, L, U;
+  std::tie(P, L, U) = at::lu_unpack(LU_data, LU_pivots);
```

Comment:
How expensive is `at::lu_unpack`? It will be called twice because `lu_solve_backward_lu_data` and `lu_solve_backward_self` are invoked separately.

I think it might be better to merge `lu_solve_backward_lu_data` and `lu_solve_backward_self` into one function and reuse temporary results like `L^{-1} P^T`.

Reply:
It is O(n^2), which one order less than any solve involved.

Reply:
Anyway, solve does the same, essentially.

Reply:
I'm also in favour of the "do less work if possible" saving the results and all that.

Reply:
Done!